### ISSUE TITLE
GHA: Sort pull requests to correct column

### ISSUE THREAD

**Issue by SAUMILDHANKAR** [hackforla/website#3112]
### Dependency
- [x] #3265

### Overview

As a developer, we have to ensure that our Kanban board is organized for all teams so that productivity is high. For this issue, we are creating a GitHub action that will change the status-column of each Pull Request (PR) based on its review progress. 

### Details
Currently, the following status-columns in our [Project Board](https://github.com/orgs/hackforla/projects/86/views/24?filterQuery=status%3A%22PR+Needs+review+%28Automated+Column%2C+do+not+place+items+here+manually%29%22%2C%22test-pending-approval+%28Automated+Column%2C+do+not+place+items+here+manually%29%22%2C%22test-approved-by-reviewer+%28Automated+Column%2C+do+not+place+items+here+manually%29%22&visibleFields=%5B%22Title%22%2C%22Assignees%22%2C%22Status%22%2C%22Milestone%22%2C%22Labels%22%5D) have been set-up such that the PRs move within these status-columns based on project automations:

#### Automated Column, do not place items here manually
- [PR Needs review](https://github.com/orgs/hackforla/projects/86/views/24?filterQuery=status%3A%22PR+Needs+review%22&visibleFields=%5B%22Title%22%2C%22Assignees%22%2C%22Status%22%2C%22Milestone%22%2C%22Labels%22%5D)
- [PRs being reviewed](https://github.com/orgs/hackforla/projects/86/views/24?filterQuery=status%3A%22PRs+being+reviewed%22)
- [PRs ✅ waiting for merge team](https://github.com/orgs/hackforla/projects/86/views/24?filterQuery=status%3A%22PRs+%E2%9C%85+waiting+for+merge+team%22)

However, this hasn't been working as desired and we have decided to use different status-column names with a GitHub action/s to sort the PRs appropriately. The new PR status-columns would be:

  - [x] PR Needs review
  - [x] PRs being reviewed
  - [ ] PRs needing changes from author 
  - [ ] PRs ready for re-review (from the original reviewer)
  - [x] PRs ✅ waiting for merge team

### Action Items
- [ ] Review how PRs currently move within different status-columns in our [Project Board](https://github.com/hackforla/website/projects/86), paying attention especially to the [Workflows](https://github.com/orgs/hackforla/projects/86/workflows) affecting PRs.
- [ ] Create/ change the status-columns suggested above in your repo. 
- [ ] Use the available Workflows and when needed create a GitHub action/s which would move the PRs among different status-columns accordingly.
- [ ] Demo your results in a team meeting.
- [ ] Once your demo is approved by the technical leads, change the names and add new new status-columns in our [Project Board](https://github.com/hackforla/website/projects/86) 
- [ ] Create a PR for the newly created GitHub action to be merged.  
- [ ] **For merge team**: Release dependency on #3227
  - [ ] Update "Action Items" in the dependency issue if needed. 
  - [ ] Remove the `Dependency` label, move the issue to the "New Issue Approval" status-column and add a `Ready for Prioritization` label.
- [ ] **For merge team**: Once the pull request associated with this issue is approved and merged, please change the name of the status-columns and/or sort the issues into the appropriate status-columns.
    - [ ] Finally delete any of the status-columns that are no longer being used.

### Pseudo Code
- [ ] _**NOTE**: This is covered by an existing Workflow, change the status-column name_. When a new PR is added to the [Project Board](https://github.com/hackforla/website/projects/86), it should automatically move to "PRs needing a reviewer assigned" status-column.
- [ ] _**NOTE**: This is covered by an existing Workflow, change the status-column name_. If the reviewer requests changes in the PR, the PR should move to "PRs needing changes from authors" column
- [ ] _**NOTE**: This will need to be addressed by a GHA:_ When PR author requests re-review from the original author, the PR should move to "PRs ready for re-review (from the original reviewer)" status-column
- [ ] _**NOTE**: Check whether the existing Workflow covers this condition, and if so change the status-column name. Otherwise address with a GHA_. If 2 reviewers have already approved the PR, the PR should move to "PRs approved, waiting for merge team to merge" status-column

### Resources/Instructions
Never done GitHub actions? [Start here!](https://docs.github.com/en/actions)
[Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
[Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
[actions/github-script](https://github.com/actions/github-script)
[GitHub RESTAPI](https://docs.github.com/en/rest)
[GitHub GraphQL](https://docs.github.com/en/graphql)
[Encrypted secrets](https://docs.github.com/en/actions/reference/encrypted-secrets)

#### Architecture Notes
- Refer to [Hack for LA's GitHub Actions (current)](https://github.com/hackforla/website/issues/6537#issuecomment-2041147335)
- Currently, we want our GitHub Actions to be structured like so based on this [proposal](https://docs.google.com/spreadsheets/d/12NcZQoyGYlHlMQtJE2IM8xLYpHN75agb/edit#gid=1231634015):

- Schedules (military time)
        - [Schedule Friday 0700](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-fri-0700.yml)
        - [Schedule Daily 1100](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-daily-1100.yml)
        - [Schedule Monthly](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-monthly.yml)
- Linters
        - [Lint SCSS](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/lint-scss.yml)
- PR Trigger
        - Add Linked Issue Labels to Pull Request
        - Add Pull Request Instructions
- Issue Trigger
        - Add Missing Labels To Issues
- WR - PR Trigger
        - WR Add Linked Issue Labels to Pull Request
        - WR Add Pull Request Instructions
- WR - Issue Trigger

Actions with the same triggers (excluding linters, which will be their own category) will live in the same GitHub action file. Scheduled actions will live in the same file if they trigger on the same schedule (i.e. all files that trigger everyday at 11am will live in one file, while files that trigger on Friday at 7am will be on a separate file).

That said, this structure is not set in stone. If any part of it feels strange, or you have questions, feel free to bring it up with the team so we can evolve this format!

**Comment by ExperimentsInHonesty** [comment]
@t-will-gillis I moved this out of the icebox.  Do you think it's ready?  If yes, add a ready for prioritization label

**Comment by HackforLABot** [comment]
Hi @t-will-gillis, thank you for taking up this issue! Hfla appreciates you :)

Do let fellow developers know about your:-
i. **Availability:** (When are you available to work on the issue/answer questions other programmers might have about your issue?)
ii. **ETA:** (When do you expect this issue to be completed?)

You're awesome!

P.S. - You may not take up another issue until this issue gets merged (or closed). Thanks again :)

**Comment by ExperimentsInHonesty** [comment]
- [ ] shorten the names of the columns that are not yet created
- [ ] create the columns
- [ ] rewrite the issue to use existing automations

**Comment by t-will-gillis** [comment]
TL; DR: I don't think this issue adds value to the project, esp. considering the complications it will bring- I feel that this issue should be closed without making these changes.

Reasoning:
- There are 3 Automated PR Columns already (we do not have a GHA workflow):
  - "PR Needs Review": Shows PR submitted, without a **_completed_** review (but might have reviewers assigned) 
  - "PRs being reviewed": At least one review of any kind has been submitted, i.e. 'Approved' or 'Correction' 
  - "PRs Waiting for Merge Team": Has at least two Approvals. 
- My argument is that having five PR columns does not address every situation and would not give you enough information to make a decision about whether or not to merge a PR, whether or not the author needs to work on the PR still, whether or not a reviewer or multiple reviewers still need to be involved, etc. 
- In order to cover every possible situation, we would need to write a GHA to cover every possible permutation which would make that GHA very complex (= very fragile), very quickly. Merge Team should be looking at each PR individually to decide what needs to be done with the PR, and not decide what to do based on the Project Board sorting.
- The reason is that having two Approvals does not automatically mean that the PR should be merged, and not having two Approvals does not automatically mean that the PR should not be merged.
- There are too many permutations of reviewers and review states (and their reviews, re-reviews, lack of re-reviews, conflicting Approval and Request Changes, Request Changes with Approval on re-review but still have open requests, and so on) to cover:
  - reviewer assigned but no review yet
  - reviewer Approves
  - reviewer Approves, but author makes changes so reviewer's review dismissed and reviewer re-Approves
  - reviewer Approves, but author makes changes so reviewer's review dismissed but reviewer disappears
  - reviewer Requests Changes
  - reviewer Requests Changes, author makes changes, reviewer's review dismissed and reviewer disappears
  - reviewer Requests Changes, author makes changes, reviewer Approves and closes change request
  - reviewer Requests Changes, author makes changes, reviewer Approves and forgets to close change request (blocking merge)
  - two (or more) reviewers Approve, last reviewer Request Changes. 


All scenarios listed above happen every week, and all require a person's judgement to decide what to do. My opinion again is that the sorting of PRs on the Project Board does not really tell us about the state of the PR, and the GHA that is required to satisfy this issue will be complex without returning value.

**Comment by t-will-gillis** [comment]
- We discussed this issue at the 6/2/25 leads meeting with BW, DA, JC, and WG. WG comments above. Also: DA and WG are using "Pull Request" page as opposed to the Project Board columns to monitor progress. 
- Closing as not planned/ won't fix.
